src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -5149,13 +5150,26 @@ protected sealed override ImmutableArray<ISymbol> GetDeclaredSymbolsCore(SyntaxN
                 return this.GetDeclaredSymbols(field, cancellationToken);
             }
 
-            var symbol = GetDeclaredSymbolCore(declaration, cancellationToken);
-            if (symbol != null)
+            // If the type decl has a primary constructor, return that symbol as well.  This is needed so that if the
+            // 'suppression' or 'generated code' attribute is on the primary constructor (i.e. by using `[method:
+            // SuppressMessage(...)]`, it will be found when walking up to the type declaration.
+            if (declaration is TypeDeclarationSyntax { ParameterList: not null } typeDeclaration)
             {
-                return ImmutableArray.Create(symbol);
+                using var result = TemporaryArray<ISymbol>.Empty;
+
+                var namedType = GetDeclaredSymbol(typeDeclaration, cancellationToken);
+                result.Add(namedType);
+
+                if (namedType.GetSymbol<NamedTypeSymbol>() is SourceMemberContainerTypeSymbol { PrimaryConstructor: { } primaryConstructor })
+                    result.Add(primaryConstructor.GetPublicSymbol());
```

Comment:
unfortunate contortions since PCs are not exposed from the public symbol model (though are understood and represented internally).  Exposing this publicly would be nice as many IDE services have also had to do special work around PCs and end up with janky code to try to figure it out.  At compiler layer this is at least a bit nicer since we can work with teh real symbols, but it's still a bit fragile as it depends on knowing that PCs are represented in precisely this fashion.